game-core/src/main/java/games/strategy/engine/framework/ui/background/TaskRunner.java

```diff
@@ -0,0 +1,96 @@
+package games.strategy.engine.framework.ui.background;
+
+import com.google.common.base.Preconditions;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Consumer;
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import javax.swing.SwingWorker;
+import javax.swing.WindowConstants;
+import lombok.Builder;
+import org.triplea.java.function.ThrowingRunnable;
+import org.triplea.swing.SwingComponents;
+
+/**
+ * Runs a background task with a wait dialog in the foreground. Similar to {@see
+ * BackgroundTaskRunner} except:
+ *
+ * <ul>
+ *   <li>Wait dialog can be closed.
+ *   <li>Does not throw exceptions, instead invokes an exception handler
+ * </ul>
+ *
+ * @param <E> Exception type that can be thrown by the background task.
+ */
+@Builder
+public class TaskRunner<E extends Exception> {
+  /**
+   * Error callback handler. The background task encounters an exception, the background task is
+   * cancelled and this is invoked with the exception that occurred.
+   */
+  @Nonnull private final Consumer<E> exceptionHandler;
+  /**
+   * Title of the wait spinner window that is displayed to user while a background action is being
+   * executed.
+   */
+  @Nonnull private final String waitDialogTitle;
+
+  /**
+   * Runs a task with a closeable wait spinner dialog in the foreground. THe wait spinner closes
+   * when the task completes. If the spinner dialog is closed, then the background task is
+   * cancelled. If there are any exceptions encountered, the spinner dialog is closed and the {@see
+   * exceptionHandler} {@code Consumer} property is called.
+   *
+   * @param backgroundAction The action to be executed.
+   */
+  public void run(final ThrowingRunnable<E> backgroundAction) {
+    Preconditions.checkNotNull(backgroundAction);
+    final AtomicReference<Throwable> exceptionRef = new AtomicReference<>();
+
+    // TODO: after https://github.com/triplea-game/triplea/pull/6001 is merged
+    //   change the 'null' parent window reference to main frame.
+    final WaitDialog waitDialog = new WaitDialog(null, waitDialogTitle);
+    waitDialog.setDefaultCloseOperation(WindowConstants.DISPOSE_ON_CLOSE);
+
+    final SwingWorker<Void, Void> worker =
+        buildBackgroundJobRunnerWithWaitDialog(backgroundAction, exceptionRef, waitDialog);
+    // If wait dialog is closed, then cancel the worker background job.
+    SwingComponents.addWindowClosedListener(waitDialog, () -> worker.cancel(true));
+
+    worker.execute();
+    waitDialog.setVisible(true);
+
+    try {
+      @SuppressWarnings("unchecked")
+      final @Nullable E exception = (E) exceptionRef.get();
+      if (exception != null) {
+        exceptionHandler.accept(exception);
+      }
+    } catch (final ClassCastException e) {
+      throw new AssertionError("Unexpected exception thrown", e);
```

Comment:
I'm 99% sure a `ClassCastException` can only ever be thrown inside the exception handler here, so in case we catch an unchecked Exception by accident causing the exceptionHandler to trip up, this exception isn't as impossible as it seems.

Thinking about it we might want to consider explicitly passing a class object here to correctly filter out the "unwanted" exception types 🤔

Reply:
> Thinking about it we might want to consider explicitly passing a class object here to correctly filter out the "unwanted" exception types thinking

This was an explicit decision. Background runner I found hard to use in the exception handling case and also turns out to be brittle (if the background task opens a swing window, because you can't close the modal wait dialog, you get a deadlock. The background task then only needs to log/throw an exception that triggers an error report dialog to crash the system). The exception API requiring an exception type was an unwelcome parameter to see, it's redundant to the generic type already. The trade-off for really specific error handling to API usage is not favorable here (otherwise background task runner could have been made fine to use).

Reply:
My only problem with this is that we only use `AssertionError` for impossible conditions (like when UTF-8 is hardcoded as charset, but an exception is thrown, or when we do a switch and we checked all possible cases so the default case can never get reached). But this particular condition is very possible in practice, so I'd prefer using `IllegalStateException`